service/worker/scanner/executions/shard/invariantManager.go

```diff
@@ -0,0 +1,125 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package shard
+
+import (
+	"github.com/uber/cadence/service/worker/scanner/executions/common"
+	"github.com/uber/cadence/service/worker/scanner/executions/invariants"
+)
+
+type (
+	invariantManager struct {
+		invariants []common.Invariant
+		types      []common.InvariantType
+	}
+)
+
+// NewInvariantManager handles running a collection of invariants according to the policy provided.
+// InvariantManager takes care of ensuring invariants are run in their correct dependency order.
+func NewInvariantManager(
```

Comment:
I don't want to overengineer this, but I have a slightly different idea about how we can implement this Policy & Manager and wanted to hear what you think about it:
1) Don't limit the manager to a single policy. Make it such that it defaults to running all policies unless a filter says otherwise. We can keep this filter in a config without hard-coding in the application.
2) Define a collection of policies (maybe we shouldn't call this policy to avoid confusion), which would act like test cases that consist of an ordered list of Invariants to test something end-to-end. Ideally, these policies should never be too long. An invariant should be able to appear in multiple policies. There shouldn't be any order/dependency requirements between policies.
3) Manager should take a list of policies to include or exclude (the config I mentioned above), and decide to run all Policies one by one, without worrying about their order. As it checks for the invariants in each policy, it can populate a map of invariant types => check results, and use that map to avoid running a given invariant repeatedly across multiple policies. At the end of the iterations, all invariants that are involved in one or more included policies will get executed exactly once.

If we follow the design above, we get the following benefits:
1) Seperation of concerns between invariants-related logic vs. the overall Scanner. Scanner interacts with a single Manager to run the tests, and the only thing it provides is the config that tells which policies should run.
2) More flexibility in the config without adding complexity. Rather than specifying a predefined policy to run, we can precisely tell what policies we want to run (and/or what policies we don't). Since the policies are not ordered or dependent on each other, the config is simple to manage. 
3) We avoid running the same invariant repeatedly if it was already covered by an earlier policy, without adding a complex invariant sorting algorithm to the mix. 

For the number of invariants & policies you currently have in the PR, my proposal above is definitely an overkill. However, if we eventually have a dozen or more invariants that are spread over multiple overlapping policies, it should scale better in terms of maintainability.